Registration Functionality

kevinliang43 opened the pull request:
# Implement Registration Functionality

## Overview
1. Registration Form Template Updates
2. Registration Form Checks and Cleaning
3. Alerts for ERROR/SUCCESS
4. User Services

## Registration Form Template Updates

### Summary
Currently, we only need the following fields to actually create a user:
1. First Name
2. Last Name
3. Username
4. Password
5. Email
6. City

I've updated the form to reflect this (Removed State and Zip code for now as it is not necessary)

We should reconsider if we really need the 'Location' table in our database.
### Screenshots
![image](https://user-images.githubusercontent.com/15353453/78398291-d53df700-75c0-11ea-8ca9-842d8bd56caf.png)

## Registration Form Checks and Cleaning

### Summary
I've added utilities to aid in checking registration submissions. These utilities can be found in the new `FormUtils.js` files. The full checks are done within the `UserService.js` file. Cleaning of the form data is done within the `UserService.js` file as well, under the `cleanFormData()` method.

### Checks
Here are the following checks being performed in the final service (Refer to `registrationChecks()` method in `UserService.js`):

1. Check to see that the `Password` and `Confirm Password` fields match.
2. Check to see that all fields of the form are not empty
3. Check to see that the given email is a valid email
4. Check to see that the chosen Username has not been taken
5. Check to see that the chosen Email is not already used.

### Cleaning
1. Remove the 'confirmPassword` field from the JSON being sent to the server.
2. Normalize the city field.

## Alerts for ERROR/SUCCESS
### Summary
Going through the client side registration process (Filling out the form, hitting the register button, and going through the checks) will return either a `SUCCESS` or an `ERROR` JSON message (`registerUser()` method). This message will either be in the form of:

```
{
    'type': 'SUCCESS',
    'newUsername': <newly registered Username>
}
```
or

```
{
    'type': 'ERROR',
    'errorMessage': <Registration Error Message>
}
```

These messages will be then passed to the RegistrationComponent to render as an alert.

### Screenshots

#### Errors
![image](https://user-images.githubusercontent.com/15353453/78417854-64b8c980-7604-11ea-932a-87ab9dceb0be.png)

#### Success
![image](https://user-images.githubusercontent.com/15353453/78399264-96a93c00-75c2-11ea-9610-06ba5304a87c.png)

## User Services

Refer to the `UserService.js` File.

### Summary
1. Made basic CRUD Services (CreateUser, findUser by ID, UpdateUser, and deleteUser)
2. More specific Read Commands (findAllUsers, findUserById, findUserByEmail, findUserByUsername)
3. registerUser -- the main method that will perform registration.

### registerUser Method

This method is the main method that the Registration Form will use to perform registrations.
Process:
1. Perform all registation checks
2. If registration checks pass, 
- clean the user data
- try to submit a `createUser` request with the cleaned user data
- if it succeeds, return a `SUCCESS` message
- if it fails, return an `ERROR` message <- this will be a Server Side registration error 
3. If registration checks fail,
- Return an `ERROR` message with the CLIENT SIDE error message.

mountainduo commented on src/services/UserService.js:
```diff
@@ -0,0 +1,143 @@
+    const usernameCheck = await findUserByUsername(user.username)
+
+    // Check to see password and confirm password match
+    if (!checkParamsMatch(user.password, user.confirmPassword)) {
+        returnJson['errorMessage'] = 'Passwords do not match.'
+    }
+
+    // Check to see that all fields are completed
+    else if (!checkForEmptyFields(user)) {
+        returnJson['errorMessage'] = 'All fields of the form must be filled out.'
+    }
+
+    //TODO: Check to see valid email <something@something.com>
+    else if (!checkEmailFormat(user.email)) {
+        returnJson['errorMessage'] = 'Invalid Email.'
+    }
+
+    // Check if email exists already
+    else if (Object.keys(emailCheck).length !== 0) {
+        returnJson['errorMessage'] = 'The Email you have entered is already being used.'
+    }
+
+    // Check if username exists already
+    else if (Object.keys(usernameCheck).length !== 0) {
+        returnJson['errorMessage'] = 'The Username you have entered is already being used.'
+    }
+
+    else {
+        returnJson['type'] = 'SUCCESS'
+    }
```
Looking good and well-organized. One point of concern might be the use of else-ifs. What if we have multiple errors? For example, if the username I want is already taken AND my passwords do not match, I would want both error messages so I can fix both mistakes.

kevinliang43 replied:
![image](https://user-images.githubusercontent.com/15353453/78417867-89ad3c80-7604-11ea-8594-bc3de3a8a6d5.png)

kevinliang43 replied:
Added a comment below with the changes

mountainduo commented on src/services/UserService.js:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    // Check if email exists already
+    else if (Object.keys(emailCheck).length !== 0) {
+        returnJson['errorMessage'] = 'The Email you have entered is already being used.'
+    }
+
+    // Check if username exists already
+    else if (Object.keys(usernameCheck).length !== 0) {
+        returnJson['errorMessage'] = 'The Username you have entered is already being used.'
+    }
```
nit: have you considered creating methods for these aspects of form validation in `FormUtils.js`?

kevinliang43 replied:
As per our discussion, lets keep it in UserService for now, as the method is specific to UserRegistration, rather than a generic method than any service could use.

mountainduo commented on src/utils/FormUtils.js:
```diff
@@ -0,0 +1,24 @@
+
+export const checkEmailFormat = (email) => {
+    // Credit: https://stackoverflow.com/questions/46155/how-to-validate-an-email-address-in-javascript
+    var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
damn wtf regex

mountainduo replied:
but nice

kevinliang43 replied:
lmao

kevinliang43 commented:
Login/Register page requirements

- [ ] Must allow users to register and create a new account

- [ ] Must allow choosing a role(s) for a user. For instance, when signing up you can provide a checkbox or radio button to select the role or roles. Alternatively provide an admin role and admin page that allows configuring user role(s)
- [ ]  Must allow login in and identifying themselves
- [ ]  Must disallow access to at least one Web page unless logged in
- [ ]  Must allow access to all other Web pages even when not logged in
- [ ] Must adapt content based on whether user is logged in or not for at least the Home page and Profile page
- [ ]  Must force login only when identity is required. For instance, an anonymous user might search for movies and visit the details page for a particular movie without needing to login. But if they attempt to like the movie, or rate it, or comment on it, or write a review, or follow someone, the application must request the user to login. Most of the Web application must be available without login (see me if not)
- [ ]  Must be mapped to /login if both login and register are implemented in the same page
The login and register page can be implemented as a single page or as two separate pages. In that case the login page must be mapped to /login and the register page must be mapped to /register

kevinliang43 commented:
New Error Alert now shows all error messages associated with any registration errors made by the User:
![image](https://user-images.githubusercontent.com/15353453/78417865-82862e80-7604-11ea-96e9-eab3c719eaf3.png)

kevinliang43 commented:
Updated Form:
![image](https://user-images.githubusercontent.com/15353453/78531694-e569f800-77b3-11ea-88ae-0652e3ec986c.png)

We are now retrieving location information dynamically from [this page on Craigslist](https://www.craigslist.org/about/sites). Look into `CraigslistService.js` for more on how this is done (More specifically, `parseRegions`).

I've removed the `City` input field and replaced it with 3 dropdown menus:
1. Continent
2. Region
3. Subregion.

These dropdown menus will update dynamically based on the preceding menu selections (Ex: If `Continent` is updated from US -> Europe, then `Region` will be populated by European Regions and `Subregion` will be populated by the subregions of the first region in the updated `Region` dropdown).

Screenshot Example:
1. Begin at (US, Alabama, auburn)
![image](https://user-images.githubusercontent.com/15353453/78531924-61644000-77b4-11ea-8356-cb0fc18ddccd.png)
2. Switch to 'Europe' in the continent dropdown, will update `Region` and `Subregion`
![image](https://user-images.githubusercontent.com/15353453/78532026-835dc280-77b4-11ea-9a97-a4b22e782dcd.png)

Infrastructure is now in place for application-wide state management, so we can begin on handling Logins after this is done.

mountainduo commented on src/components/register/RegisterComponent.js:
```diff
@@ -19,121 +124,129 @@ const RegisterComponent = () =>
-                    <button className="btn btn-block btn-success">Register</button>
+                    <button className="btn btn-block btn-success"
+                        onClick={(e) => {
+                            e.preventDefault();//FIXME: Temporary workaround Prevents refreshing of the page (in order to show alert.)
```
Ah you might've explained this to me in our last call, but do you have a plan set for a better way to get the error messages to show up? either way, any way i can help?

mountainduo commented on src/components/register/RegisterComponent.js:
```diff
@@ -19,121 +124,129 @@ const RegisterComponent = () =>
+                        <label for="inputContinent" class="col-form-label text-white float-left">Continent</label>
+                        <select id="inputContinent" class="form-control"
+                            onChange={(e) => {
+                                let selectedContinent = e.target.value;
+                                let regionList = this.getRegionsForContinent(selectedContinent);
+                                let selectedRegion = regionList[0];
+                                let subregionList = this.getSubregionsForRegion(selectedContinent, selectedRegion);
+                                let selectedSubregion = subregionList[0][1];
+                    
+                                this.setState(prevState => (
+                                    {
+                                        ...prevState,
+                                        'selectedContinent': selectedContinent,
+                                        'regionList': regionList,
+                                        'selectedRegion': selectedRegion,
+                                        'subregionList': subregionList,
+                                        'selectedSubregion': selectedSubregion
+                                    }))}}>
```
i like it! i think this is great as it is right now, but it'd also be nice to have the default values for Continent, Region, and Subregion be `Choose` or `---`, rather than US, Alabama, and Auburn, respectively. This would require a new type of error message for form validation failure when Continent, Region, and Subregion aren't correctly selected though. Time permitting, i'd be happy to go ahead and try to add this.

kevinliang43 commented on src/services/UserService.js:
```diff
@@ -0,0 +1,146 @@
+
+    // Check to see password and confirm password match
+    if (!checkParamsMatch(user.password, user.confirmPassword)) {
+        returnJson['errorMessages'].push('Passwords do not match.')
```
Yep lets keep a note for this and add this in a future PR.
